crates/telio-batcher/src/batcher.rs

```diff
@@ -0,0 +1,404 @@
+use futures::future::BoxFuture;
+use std::fmt::Debug;
+use std::hash::Hash;
+use std::{collections::HashMap, sync::Arc};
+
+use telio_utils::telio_log_debug;
+use tokio::time::sleep_until;
+
+type Action<V, R = std::result::Result<(), ()>> =
+    Arc<dyn for<'a> Fn(&'a mut V) -> BoxFuture<'a, R> + Sync + Send>;
+
+pub struct Batcher<K, V> {
+    actions: HashMap<K, (BatchEntry, Action<V>)>,
+}
+
+struct BatchEntry {
+    deadline: tokio::time::Instant,
+    interval: std::time::Duration,
+    threshold: std::time::Duration,
+}
+
+impl<K, V> Default for Batcher<K, V>
+where
+    K: Eq + Hash + Send + Sync + Debug + Clone,
+{
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<K, V> Batcher<K, V>
+where
+    K: Eq + Hash + Send + Sync + Debug + Clone,
+{
+    pub fn new() -> Self {
+        Self {
+            actions: HashMap::new(),
+        }
+    }
+
+    /// Batching works by sleeping until the nearest future and then trying to batch more actions
+    /// based on the threshold value. Higher delay before calling the function will increase the chances of batching
+    /// because the deadlines will _probably_ be in the past already.
+    pub async fn get_actions(&mut self) -> Option<Vec<(K, Action<V>)>> {
```

Comment:
What is the benefit for having both None and vec![], here?
Maybe just return empty vector, if there are no actions to do

Reply:
You're totally right - an empty vector and None are basically the same thing. However from my experiments SessionKeeper tried really hard calling it's `wait_with_update()` issued by the `Task` when I resolved immediately that I decided to do something about it.

To be honest I don't feel like this is the best solution except for the fact I can do `Some(_) = ... => ` inside of tokio::select!  :)

Reply:
Then it should mostly be hidden area, there you are entering else pending or somthing of the like, you can always add pending on 0

Reply:
@tomaszklak @packgron I've added a notifier into the `get_actions()`.

It also should handle the case where adding an action to batcher would delay if already awaiting, this should be handled now as well.

Reply:
There's one caveat though. Given that Batcher is under a `Mutex` it will block for however long it is needed, naturally, so for example `get_actions()` will block until there's one item which means it will hold the mutex and will not allow it to be released for adding the item, in order for this to break, the call site needs to have a  tokio::select or tokio::timeout or some other form of _retry_.

Currently SessionKeeper and RepeatedTask doesn't have a mutex and RepeatedTask blocks until one of it's features resolves, meaning it would also block. However IMHO from my research the SessionKeeper's `wait_with_update()` will have another tokio::select! branch triggered which basically _fixes_ the blocking.

Not sure yet how relevant this is for Batcher though.

Here's a simple unit-test that fails at the moment:
```
#[tokio::test(start_paused = true)]
    async fn batch_blocks_until_theres_an_item() {
        let start_time = tokio::time::Instant::now();
        let mut batcher = Arc::new(Mutex::new(Batcher::<String, TestChecker>::new()));

        let batcher2 = batcher.clone();
        let handle = tokio::spawn(async move {
            tokio::time::advance(time::Duration::from_secs(50)).await;

            batcher2.lock().await.add(
                "key0".to_owned(),
                Duration::from_secs(100),
                Duration::from_secs(0),
                Arc::new(|s: _| {
                    Box::pin(async move {
                        s.values
                            .push(("key0".to_owned(), tokio::time::Instant::now()));
                        Ok(())
                    })
                }),
            );
        });

        
        let mut batcher = batcher.lock().await;
        let actions_fut = batcher.get_actions();
        let actions = tokio::time::timeout(time::Duration::from_secs(100), actions_fut).await;
        
        assert!(actions.is_ok());
        assert!(actions.unwrap().len() == 1);
    }
```


However this problem is basically not related to Batcher but in general is a problem in such situations. @tomaszklak @packgron opinions on this?

Reply:
So there is 3 ways two think about this IMO:
- Add's are done in built faze, and repeats in wait step, so maybe we could just make api reflect that (I don't know if this is 100% true though). If we are sure those are two idepenedent actions we don't need to wait
- We need runtime additions / dealetions, in this case we could use tokio::sync::watch, instead of Mutex, and react to change in configuration.
- Leave it as small pitfall, and rely it's aways used in select arm

Reply:
Okay. I will leave it as it is. This is because:
- I am not yet clear how it's gonna be used(so far only in direct connection cases through SessionKeeper)
- Batcher is under a feature flag
- Even though it never resolves on `get_actions()` if it's empty and nothing is added it will introduce no blockage into current libtelio workflow.